File: FCL/src/main/java/com/tungsten/fcl/ui/controller/EditableControllerListAdapter.java
package com.tungsten.fcl.ui.controller;

import android.annotation.SuppressLint;
import android.content.Context;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import androidx.constraintlayout.widget.ConstraintLayout;

import com.tungsten.fcl.R;
import com.tungsten.fcl.setting.Controller;
import com.tungsten.fcl.ui.UIManager;
import com.tungsten.fclcore.util.flow.FlowSubscriptions;
import com.tungsten.fcllibrary.component.FCLAdapter;
import com.tungsten.fcllibrary.component.dialog.FCLAlertDialog;
import com.tungsten.fcllibrary.component.view.FCLImageButton;
import com.tungsten.fcllibrary.component.view.FCLTextView;

import java.util.List;

public class EditableControllerListAdapter extends FCLAdapter {

    private final List<Controller> list;

    public EditableControllerListAdapter(Context context, List<Controller> list) {
        super(context);
        this.list = list;
    }

    static class ViewHolder {
        ConstraintLayout parent;
        FCLTextView name;
        FCLTextView version;
        FCLImageButton delete;
        // 阶段 4a：Controller 属性已 StateFlow 化；视图回收重绑前取消旧订阅
        //（对齐原 bind 重复调用先解绑的语义）。
        FlowSubscriptions.Subscription nameSubscription;
        FlowSubscriptions.Subscription versionSubscription;
    }

    @Override
    public int getCount() {
        return list.size();
    }

    @Override
    public Object getItem(int i) {
        return list.get(i);
    }

    @SuppressLint("UseCompatLoadingForDrawables")
    @Override
    public View getView(int i, View view, ViewGroup viewGroup) {
        final ViewHolder viewHolder;
        if (view == null) {
            viewHolder = new ViewHolder();
            view = LayoutInflater.from(getContext()).inflate(R.layout.item_controller_editable, null);
            viewHolder.parent = view.findViewById(R.id.parent);
            viewHolder.name = view.findViewById(R.id.name);
            viewHolder.version = view.findViewById(R.id.version);
            viewHolder.delete = view.findViewById(R.id.delete);
            view.setTag(viewHolder);
        } else {
            viewHolder = (ViewHolder) view.getTag();
        }
        Controller controller = list.get(i);
        viewHolder.parent.setBackground(controller == ((ControllerManagePage) UIManager.getInstance().getControllerUI().getPage(ControllerPageManager.PAGE_ID_CONTROLLER_MANAGER)).getSelectedController() ? getContext().getDrawable(R.drawable.bg_container_transparent_selected) : getContext().getDrawable(R.drawable.bg_container_transparent_clickable));
        if (viewHolder.nameSubscription != null)
            viewHolder.nameSubscription.cancel();
        viewHolder.nameSubscription = FlowSubscriptions.subscribeWithCurrent(controller.nameFlow(), v -> viewHolder.name.stringProperty().setValue(v));
        if (viewHolder.versionSubscription != null)
            viewHolder.versionSubscription.cancel();
        viewHolder.versionSubscription = FlowSubscriptions.subscribeWithCurrent(controller.versionFlow(), v -> viewHolder.version.stringProperty().setValue(v));
        viewHolder.parent.setOnClickListener(view1 -> {
            ((ControllerManagePage) UIManager.getInstance().getControllerUI().getPage(ControllerPageManager.PAGE_ID_CONTROLLER_MANAGER)).setSelectedController(controller);
            notifyDataSetChanged();
        });
        viewHolder.delete.setOnClickListener(view1 -> {
            FCLAlertDialog.Builder builder = new FCLAlertDialog.Builder(getContext());
            builder.setAlertLevel(FCLAlertDialog.AlertLevel.INFO);
            builder.setCancelable(false);
            builder.setMessage(getContext().getString(R.string.control_delete));
            builder.setPositiveButton(() -> ((ControllerManagePage) UIManager.getInstance().getControllerUI().getPage(ControllerPageManager.PAGE_ID_CONTROLLER_MANAGER)).removeController(controller));
            builder.setNegativeButton(null);
            builder.create().show();
        });
        return view;
    }
}

File: FCL/src/main/java/com/tungsten/fcl/ui/controller/EditableControllerListAdapter.kt
package com.tungsten.fcl.ui.controller

import android.annotation.SuppressLint
import android.content.Context
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.constraintlayout.widget.ConstraintLayout
import com.tungsten.fcl.R
import com.tungsten.fcl.setting.Controller
import com.tungsten.fcllibrary.component.FCLAdapter
import com.tungsten.fcllibrary.component.dialog.FCLAlertDialog
import com.tungsten.fcllibrary.component.view.FCLImageButton
import com.tungsten.fcllibrary.component.view.FCLTextView

/**
 * 控制布局列表适配器：列表项选择/删除通过回调交给页面处理，
 * 选中高亮由 [getSelected] 提供，不使用 fakefx property。
 */
class EditableControllerListAdapter(
    context: Context,
    private val list: List<Controller>,
    private val getSelected: () -> Controller?,
    private val onSelect: (Controller) -> Unit,
    private val onDelete: (Controller) -> Unit
) : FCLAdapter(context) {

    override fun getCount(): Int = list.size

    override fun getItem(i: Int): Any = list[i]

    @SuppressLint("UseCompatLoadingForDrawables")
    override fun getView(i: Int, convertView: View?, parent: ViewGroup): View {
        val view: View
        val holder: ViewHolder
        if (convertView == null) {
            holder = ViewHolder()
            view = LayoutInflater.from(context)
                .inflate(R.layout.item_controller_editable, parent, false)
            holder.parent = view.findViewById(R.id.parent)
            holder.name = view.findViewById(R.id.name)
            holder.version = view.findViewById(R.id.version)
            holder.delete = view.findViewById(R.id.delete)
            view.tag = holder
        } else {
            view = convertView
            holder = view.tag as ViewHolder
        }
        val controller = list[i]
        holder.parent.background = if (controller == getSelected()) {
            context.getDrawable(R.drawable.bg_container_transparent_selected)
        } else {
            context.getDrawable(R.drawable.bg_container_transparent_clickable)
        }
        holder.name.text = controller.name
        holder.version.text = controller.version
        holder.parent.setOnClickListener { onSelect(controller) }
        holder.delete.setOnClickListener {
            FCLAlertDialog.Builder(context)
                .setAlertLevel(FCLAlertDialog.AlertLevel.INFO)
                .setCancelable(false)
                .setMessage(context.getString(R.string.control_delete))
                .setPositiveButton { onDelete(controller) }
                .setNegativeButton(null)
                .create()
                .show()
        }
        return view
    }

    private class ViewHolder {
        lateinit var parent: ConstraintLayout
        lateinit var name: FCLTextView
        lateinit var version: FCLTextView
        lateinit var delete: FCLImageButton
    }
}
